File: examples/android/src/main/java/com/dropbox/core/examples/android/DownloadFileTask.java
package com.dropbox.core.examples.android;

import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Environment;

import com.dropbox.core.DbxException;
import com.dropbox.core.v2.DbxFiles;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;

/**
 * Task to download a file from Dropbox and put it in the Downloads folder
 */
class DownloadFileTask extends AsyncTask<DbxFiles.FileMetadata, Void, File> {

    private final Context mContext;
    private final DbxFiles mFilesClient;
    private Exception mException;
    private Callback mCallback;

    public interface Callback {
        void onDownloadComplete(File result);
        void onError(Exception e);
    }

    DownloadFileTask(Context context, DbxFiles filesClient, Callback callback) {
        mContext = context;
        mFilesClient = filesClient;
        mCallback = callback;
    }

    @Override
    protected void onPostExecute(File result) {
        super.onPostExecute(result);
        if (mException != null) {
            mCallback.onError(mException);
        } else {
            mCallback.onDownloadComplete(result);
        }
    }

    @Override
    protected File doInBackground(DbxFiles.FileMetadata... params) {
        DbxFiles.FileMetadata metadata = params[0];
        try {
            File path = Environment.getExternalStoragePublicDirectory(
                    Environment.DIRECTORY_DOWNLOADS);
            File file = new File(path, metadata.name);

            // Make sure the Downloads directory exists.
            path.mkdirs();

            // Upload the file.
            OutputStream outputStream = new FileOutputStream(file);
            try {
                mFilesClient.downloadBuilder(metadata.pathLower).
                        rev(metadata.rev).run(outputStream);
            } finally {
                outputStream.close();
            }

            // Tell android about the file
            Intent intent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);
            intent.setData(Uri.fromFile(file));
            mContext.sendBroadcast(intent);

            return file;
        } catch (DbxException | IOException e) {
            e.printStackTrace();
            mException = e;
        }

        return null;
    }
}

File: examples/android/src/main/java/com/dropbox/core/examples/android/DownloadFileTask.kt
package com.dropbox.core.examples.android

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Environment
import com.dropbox.core.DbxException
import com.dropbox.core.v2.DbxClientV2
import com.dropbox.core.v2.files.FileMetadata
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import kotlin.coroutines.CoroutineContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

sealed interface DownloadFileTaskResult {
    class Success(val result: File) : DownloadFileTaskResult
    class Error(val e: Exception) : DownloadFileTaskResult
}

/**
 * Task to download a file from Dropbox and put it in the Downloads folder
 */
internal class DownloadFileTask(
    private val mContext: Context,
    private val mDbxClient: DbxClientV2,
    private val dispatcher: CoroutineContext,
) {
    suspend fun download(metadata: FileMetadata): DownloadFileTaskResult = withContext(dispatcher){
         try {
            val path = Environment.getExternalStoragePublicDirectory(
                Environment.DIRECTORY_DOWNLOADS
            )
            val file = File(path, metadata.name)

            // Make sure the Downloads directory exists.
            if (!path.exists()) {
                if (!path.mkdirs()) {
                    return@withContext DownloadFileTaskResult.Error(
                        RuntimeException("Unable to create directory: $path")
                    )
                }
            } else if (!path.isDirectory) {
                return@withContext  DownloadFileTaskResult.Error(
                    IllegalStateException("Download path is not a directory: $path")
                )
            }
            FileOutputStream(file).use { outputStream ->
                mDbxClient.files().download(metadata.pathLower, metadata.rev)
                    .download(outputStream)
            }

            // Tell android about the file
            val intent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE)
            intent.data = Uri.fromFile(file)
            mContext.sendBroadcast(intent)
             DownloadFileTaskResult.Success(file)
        } catch (e: DbxException) {
             DownloadFileTaskResult.Error(e)
        } catch (e: IOException) {
             DownloadFileTaskResult.Error(e)
        }
    }
}